### PR TITLE
Add ability to copy marked text to the clipboard

### PR DESCRIPTION
**Pull request by sasumner**
Implements #6095

This one may be a bit controversial (perhaps why I haven't finished it before now), but the need keeps coming up on the Community site.
Basically, it is a user need to copy (red)marked text to the clipboard.
It really is only useful after doing regular-expression based marking via the _Mark_ dialog.

So here is where it would appear in the menus (is there a better place?):

image

Here's an example of the paste result after running the _Marked Text to Clipboard_, after marking as indicated by the search shown:

image

When the search matches run across line-endings, some sort of delimiter is needed to show where one match ends and another begins.  I chose `----` for this in the output, but it could be something else if that is not liked:

image

### CODE REVIEW

**Comment by guy038** [comment]
Hi, @sasumner, and **All**,

First, **many thanks** for your implication in the improvement of N++ and, especially, for everything related to the "Find" window tabs. You are going to become "The" specialist of these 4 tabs ;-))

Your idea of adding the possibility of copying **all marked** occurrences of a search in the **clipboard** is really valuable. And the future location of this feature ( **`Edit > Copy to Clipboard > Marked text to Clipboard`** is quite **logical**, too.

I've thought of an **other** location : **Search > Bookmark > Copy Bookmarked Text`**. Note that this **new** syntax is different from the existent **Search > Bookmark > Copy Bookmarked Lines`**

In this way, this new option could be **easily** reached by **right**-clicking on the **Bookmark** margin, that you lately implemented ;-))

What do you think of ? May be, code or structure **problems** would arise, choosing my solution ?

As usual, **Scott**, just a suggestion !

Best Regards,

guy038

**Comment by guy038** [comment]
Hello, @sasumner, and **All**,

As, apparently, I do **not** have, anymore, the possiblity of editing a **previous** post, here is an **updated** version that I hope will be more readable !

---

First, **many thanks** for your implication in the improvement of N++ and, especially, for everything related to the **"Find"** window tabs. You are going to become "The" specialist of these **4** tabs ;-))

Your idea of adding the possibility of copying **all marked** occurrences of a search, in the **clipboard**, is really valuable. And the **future** location of this feature ( **`Edit > Copy to Clipboard > Marked text to Clipboard`** ) is quite **logical**, too.

I've thought of an **other** location : **`Search > Bookmark > Copy Bookmarked Text`**. Note that this **new** syntax is different from the **existent** **`Search > Bookmark > Copy Bookmarked Lines`** option !

In this way, this **new** option could be **easily** reached by **right**-clicking on the **Bookmark** margin, that you lately implemented ;-))

What do you think of ? May be, code or structure **problems** would arise, choosing my solution ?

As usual, **Scott**, just a suggestion !

Best Regards,

guy038

**Comment by sasumner** [comment]
>many thanks ... especially, for everything related to the "Find" window tabs. You are going to become "The" specialist of these 4 tabs

It DOES seem like many of my efforts are concentrated here!

BTW, did you notice a recent change I did allows a user to reuse the same Alt+letter keyboard accelerators in the four tabs?  So, you could have Alt+r do replace in the _Replace_ tab, and Alt+r do replace-in-files in the _Find in Files_ tab (although that is not a great example!).  Until this change, I was using your long ago remapping so that every function in the 4 tabs had a unique accelerator -- remember that one? :-)

>this new option could be easily reached by right-clicking on the Bookmark margin

I like it.  I will add it.  I suppose there is no reason that it shouldn't appear in two places in the UI.

**Comment by sasumner** [comment]
>I like it. I will add it. I suppose there is no reason that it shouldn't appear in two places in the UI.

I thought some more about it.
I will add it as an enhancement after this PR is merged (if it is).

**Comment by donho** [comment]
`Edit->Copy to Clipboard` is a quite logical choice. However, why not a button "Marked Text to Clipboard" just under "Mark All" in Find dialog's Mark section? It will more easy to use and understand.

**Comment by sasumner** [comment]
>why not a button "Marked Text to Clipboard" just under "Mark All" in Find dialog's Mark section?

Sure, we can do that.  That text is kind of long for a button; could be multiline I suppose...

**Review by donho — CHANGES_REQUESTED** [review]
1. Add a button "Marked text to Clipboard".
2. Add a generic méthod `markedToClipboard(int indiStyle)` for the future commands "Marked Style N text to Clipboard".

**Comment by sasumner** [comment]
I shortened the text for the button a little bit so that it doesn't need to be multiline.
I tried it as a multiline button, but the reality of it was that it was so big on the screen that way that it looked like a very important feature (really this is a minor thing).
I left the menu entry in the _Edit_ menu because that way the command is macroable without having to be supported in the complicated way that other find-related macros get recorded.

File: PowerEditor/src/Notepad_plus.rc [hunk 1]
```diff
@@ -283,6 +283,8 @@ BEGIN
                 MENUITEM "Current Full File path to Clipboard",    IDM_EDIT_FULLPATHTOCLIP
                 MENUITEM "Current Filename to Clipboard",          IDM_EDIT_FILENAMETOCLIP
                 MENUITEM "Current Dir. Path to Clipboard",         IDM_EDIT_CURRENTDIRTOCLIP
+                MENUITEM SEPARATOR
+                MENUITEM "Marked Text to Clipboard",               IDM_EDIT_MARKEDTOCLIP
```
Comment (donho) [inline PowerEditor/src/Notepad_plus.rc:287]: IMO, it's not necessary - that's why the existence of the button.

Comment (sasumner) [reply PowerEditor/src/Notepad_plus.rc:287]: Did you see my comment https://github.com/notepad-plus-plus/notepad-plus-plus/pull/8867#issuecomment-700976991 about macro-ability?

Do we leave it unable to be recorded in a macro?

Or do we give it this type of treatment: https://npp-user-manual.org/docs/searching/#searching-actions-when-recorded-as-macros
This type seems like a lot of effort for such a simple command.

Comment (donho) [reply PowerEditor/src/Notepad_plus.rc:287]: OK, it's not so bad to have menu command.
We'll have "Marked Style N text to Clipboard" (N:1..5) &  "All Marked text to Clipboard" 6 commands:
https://github.com/notepad-plus-plus/notepad-plus-plus/issues/8943
We may make sure the name "Marked Text to Clipboard" is not ambiguous among these commands.

Comment (donho) [reply PowerEditor/src/Notepad_plus.rc:287]: BTW, feel free to assign to yourself if you're interested in doing https://github.com/notepad-plus-plus/notepad-plus-plus/issues/8943

File: PowerEditor/src/ScitillaComponent/FindReplaceDlg.rc [hunk 2]
```diff
@@ -80,6 +80,7 @@ BEGIN
     PUSHBUTTON      "Replace All in All Opened Doc&uments",IDC_REPLACE_OPENEDFILES,268,74,91,21,BS_MULTILINE
     PUSHBUTTON      "Find All",IDD_FINDINFILES_FIND_BUTTON,268,20,91,14,WS_GROUP
     PUSHBUTTON      "Replace in Files",IDD_FINDINFILES_REPLACEINFILES,268,38,91,14
+    PUSHBUTTON      "Copy Marked Text",IDC_COPY_MARKED_TEXT,268,56,91,14
```
Comment (donho) [inline PowerEditor/src/ScitillaComponent/FindReplaceDlg.rc:83]: Nice one :)

File: PowerEditor/src/Notepad_plus.rc (hunk 1 above)
Comment (sasumner) [reply PowerEditor/src/Notepad_plus.rc:287]: OK, I will do it, once this one is committed; that way there shouldn't be any conflicts to resolve around MENUITEM in Notepad_plus.rc.

**Comment by sasumner** [comment]
@guy038 said:

>I've thought of an other location : Search > Bookmark > Copy Bookmarked Text.
>In this way, this new option could be easily reached by right-clicking on the Bookmark margin, that you lately implemented

I think this "need" goes away with @donho 's suggestion of adding a button to the _Mark_ window, so I won't do it. :-)

File: PowerEditor/src/Notepad_plus.rc (hunk 1 above)
Comment (sasumner) [reply PowerEditor/src/Notepad_plus.rc:287]: >make sure the name "Marked Text to Clipboard" is not ambiguous 

Like the existing _Search_ > _Jump up_ > ... commands, we can use something like _Copy Style #**_N_** Text to Clipboard_

File: PowerEditor/src/ScitillaComponent/FindReplaceDlg.cpp [hunk 3]
```diff
@@ -3390,6 +3404,48 @@ bool FindReplaceDlg::replaceInOpenDocsConfirmCheck(void)
 	return confirmed;
 }
 
+void FindReplaceDlg::markedTextToClipboard(int indiStyle)
```
Comment (donho) [inline PowerEditor/src/ScitillaComponent/FindReplaceDlg.cpp:3407]: According https://github.com/notepad-plus-plus/notepad-plus-plus/issues/8943 ,  `markedTextToClipboard` method shouldn't belong to `FindReplaceDlg`

**Review by donho — CHANGES_REQUESTED** [review]
Find a suitable class for `markedTextToClipboard` method.

File: PowerEditor/src/ScitillaComponent/FindReplaceDlg.cpp (hunk 3 above)
Comment (sasumner) [reply PowerEditor/src/ScitillaComponent/FindReplaceDlg.cpp:3407]: Haha. Future issues affect current issues.

Comment (donho) [reply PowerEditor/src/ScitillaComponent/FindReplaceDlg.cpp:3407]: > Haha. Future issues affect current issues.

Not really, markedTextToClipboard has been made more generic, so it should be in a place more generic :)

Comment (sasumner) [reply PowerEditor/src/ScitillaComponent/FindReplaceDlg.cpp:3407]: Probably to do it well, we should cancel this PR, combine the two issues into one, and have a cohesive approach.
I'm usually in favor of separating things, but in this case maybe together is better.
If you agree, close this PR?

Comment (donho) [reply PowerEditor/src/ScitillaComponent/FindReplaceDlg.cpp:3407]: Do you prefer to move markedTextToClipboard in the PR for #8943 ?

Comment (sasumner) [reply PowerEditor/src/ScitillaComponent/FindReplaceDlg.cpp:3407]: Short answer: Yes. :-)

**Pull request closed without merging**